File: packages/go/storage/mysql/mysql.go
package mysql

import (
	"context"
	"database/sql"
	"encoding/hex"
	"errors"
	"fmt"
	"strings"
	"time"

	mysqlDriver "github.com/go-sql-driver/mysql"
	identifold "github.com/greyfoundry/identifold/packages/go"
	"github.com/greyfoundry/identifold/packages/go/storage"
)

type queryer interface {
	QueryContext(context.Context, string, ...any) (*sql.Rows, error)
	QueryRowContext(context.Context, string, ...any) *sql.Row
}

// Adapter uses a caller-owned database/sql handle backed by MySQL or MariaDB.
type Adapter struct {
	database queryer
}

func New(database queryer) *Adapter {
	return &Adapter{database: database}
}

func (adapter *Adapter) Reserve(ctx context.Context, request storage.ReferenceReservation) (bool, error) {
	machineID, err := machineIDBytes(request.MachineID)
	if err != nil {
		return false, err
	}
	rows, err := adapter.database.QueryContext(
		ctx,
		"CALL identifold_reserve_reference(?, ?, ?)",
		machineID,
		request.Namespace,
		request.Reference,
	)
	if err != nil {
		return false, mapError(err)
	}
	defer rows.Close()
	if !rows.Next() {
		return false, conflict()
	}
	var reserved bool
	if err := rows.Scan(&reserved); err != nil || rows.Next() {
		return false, conflict()
	}
	return reserved, nil
}

func (adapter *Adapter) Resolve(ctx context.Context, reference string, namespace string) (*storage.ReferenceMapping, error) {
	mapping, err := adapter.lookup(
		ctx,
		"SELECT machine_id, namespace FROM identifold_references WHERE reference = ? AND namespace = ?",
		reference,
		namespace,
	)
	if err != nil || mapping != nil {
		return mapping, err
	}
	prefix, scope, sequence, ok := parseSequentialReference(reference)
	if !ok {
		return nil, nil
	}
	return adapter.lookup(
		ctx,
		"SELECT machine_id, namespace FROM identifold_sequence_allocations WHERE namespace = ? AND reference_prefix = ? AND scope = ? AND sequence = ?",
		namespace,
		prefix,
		scope,
		sequence,
	)
}

func (adapter *Adapter) Allocate(ctx context.Context, request storage.SequenceAllocationRequest) (uint64, error) {
	machineID, err := machineIDBytes(request.MachineID)
	if err != nil {
		return 0, err
	}
	for attempt := 0; attempt < 5; attempt++ {
		rows, queryErr := adapter.database.QueryContext(
			ctx,
			"CALL identifold_allocate_sequence(?, ?, ?, ?, ?)",
			machineID,
			request.Namespace,
			request.ReferencePrefix,
			request.Scope,
			request.Width,
		)
		if queryErr == nil {
			sequence, scanErr := scanSequence(rows)
			if scanErr == nil {
				return sequence, nil
			}
			queryErr = scanErr
		}
		if attempt == 4 || !isTransient(queryErr) {
			return 0, mapError(queryErr)
		}
		select {
		case <-ctx.Done():
			return 0, conflict()
		case <-time.After(time.Duration(1<<attempt) * time.Millisecond):
		}
	}
	return 0, conflict()
}

func (adapter *Adapter) lookup(ctx context.Context, query string, arguments ...any) (*storage.ReferenceMapping, error) {
	var machineID []byte
	var namespace string
	err := adapter.database.QueryRowContext(ctx, query, arguments...).Scan(&machineID, &namespace)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, mapError(err)
	}
	value, err := machineIDString(machineID)
	if err != nil {
		return nil, err
	}
	return &storage.ReferenceMapping{MachineID: value, Namespace: namespace}, nil
}

func scanSequence(rows *sql.Rows) (uint64, error) {
	defer rows.Close()
	if !rows.Next() {
		return 0, conflict()
	}
	var sequence uint64
	if err := rows.Scan(&sequence); err != nil || rows.Next() {
		return 0, conflict()
	}
	return sequence, nil
}

func machineIDBytes(value string) ([]byte, error) {
	hexadecimal := strings.ReplaceAll(value, "-", "")
	if len(hexadecimal) != 32 {
		return nil, conflict()
	}
	result, err := hex.DecodeString(hexadecimal)
	if err != nil {
		return nil, conflict()
	}
	return result, nil
}

func machineIDString(value []byte) (string, error) {
	if len(value) != 16 {
		return "", conflict()
	}
	hexadecimal := hex.EncodeToString(value)
	return fmt.Sprintf(
		"%s-%s-%s-%s-%s",
		hexadecimal[0:8],
		hexadecimal[8:12],
		hexadecimal[12:16],
		hexadecimal[16:20],
		hexadecimal[20:32],
	), nil
}

func parseSequentialReference(reference string) (string, string, string, bool) {
	parts := strings.Split(reference, "-")
	var prefix, scope, sequence, check string
	switch len(parts) {
	case 3:
		prefix, sequence, check = parts[0], parts[1], parts[2]
	case 4:
		prefix, scope, sequence, check = parts[0], parts[1], parts[2], parts[3]
	default:
		return "", "", "", false
	}
	if len(prefix) < 2 || len(prefix) > 8 || !asciiRange(prefix, 'A', 'Z') ||
		(scope != "" && (len(scope) != 4 || !asciiRange(scope, '0', '9'))) ||
		len(sequence) < 4 || len(sequence) > 18 || !asciiRange(sequence, '0', '9') ||
		len(check) != 1 {
		return "", "", "", false
	}
	return prefix, scope, sequence, true
}

func asciiRange(value string, minimum byte, maximum byte) bool {
	for index := range len(value) {
		if value[index] < minimum || value[index] > maximum {
			return false
		}
	}
	return true
}

func mapError(err error) error {
	var databaseError *mysqlDriver.MySQLError
	if errors.As(err, &databaseError) {
		switch string(databaseError.SQLState[:]) {
		case "22003":
			return &identifold.Error{Code: "sequence_overflow"}
		case "22023":
			return &identifold.Error{Code: "invalid_allocation_policy"}
		}
	}
	return conflict()
}

func isTransient(err error) bool {
	var databaseError *mysqlDriver.MySQLError
	return errors.As(err, &databaseError) &&
		(databaseError.Number == 1205 || databaseError.Number == 1213 ||
			string(databaseError.SQLState[:]) == "40001")
}

func conflict() error {
	return &identifold.Error{Code: "allocation_conflict"}
}

var _ storage.Adapter = (*Adapter)(nil)

File: packages/rust/src/storage/mysql.rs
use std::time::Duration;

use mysql_async::{Error as MySqlError, Pool, params, prelude::Queryable};

use super::{
    ReferenceMapping, ReferenceReservation, SequenceAllocationRequest, StorageAdapter,
    StorageFuture,
};
use crate::Error;

pub struct MySqlStorageAdapter {
    pool: Pool,
}

impl MySqlStorageAdapter {
    pub fn new(pool: Pool) -> Self {
        Self { pool }
    }
}

impl StorageAdapter for MySqlStorageAdapter {
    fn reserve<'a>(&'a self, request: &'a ReferenceReservation) -> StorageFuture<'a, bool> {
        Box::pin(async move {
            let mut connection = self.pool.get_conn().await.map_err(map_error)?;
            let row: Option<(u8,)> = connection
                .exec_first(
                    "CALL identifold_reserve_reference(:machine_id, :namespace, :reference)",
                    params! {
                        "machine_id" => machine_id_bytes(&request.machine_id)?,
                        "namespace" => request.namespace.as_str(),
                        "reference" => request.reference.as_str(),
                    },
                )
                .await
                .map_err(map_error)?;
            match row {
                Some((0,)) => Ok(false),
                Some((1,)) => Ok(true),
                _ => Err(conflict()),
            }
        })
    }

    fn resolve<'a>(
        &'a self,
        reference: &'a str,
        namespace: &'a str,
    ) -> StorageFuture<'a, Option<ReferenceMapping>> {
        Box::pin(async move {
            let mut connection = self.pool.get_conn().await.map_err(map_error)?;
            let random: Option<(Vec<u8>, String)> = connection
                .exec_first(
                    "SELECT machine_id, namespace FROM identifold_references \
                     WHERE reference = :reference AND namespace = :namespace",
                    params! {
                        "reference" => reference,
                        "namespace" => namespace,
                    },
                )
                .await
                .map_err(map_error)?;
            if let Some((machine_id, namespace)) = random {
                return Ok(Some(mapping(machine_id, namespace)?));
            }

            let Some((prefix, scope, sequence)) = parse_sequential_reference(reference) else {
                return Ok(None);
            };
            let row: Option<(Vec<u8>, String)> = connection
                .exec_first(
                    "SELECT machine_id, namespace FROM identifold_sequence_allocations \
                     WHERE namespace = :namespace AND reference_prefix = :prefix \
                     AND scope = :scope AND sequence = :sequence",
                    params! {
                        "namespace" => namespace,
                        "prefix" => prefix,
                        "scope" => scope,
                        "sequence" => sequence,
                    },
                )
                .await
                .map_err(map_error)?;
            row.map(|(machine_id, namespace)| mapping(machine_id, namespace))
                .transpose()
        })
    }

    fn allocate<'a>(&'a self, request: &'a SequenceAllocationRequest) -> StorageFuture<'a, u64> {
        Box::pin(async move {
            let machine_id = machine_id_bytes(&request.machine_id)?;
            for attempt in 0..5 {
                let result = allocate_once(&self.pool, request, &machine_id).await;
                match result {
                    Ok(sequence) => return Ok(sequence),
                    Err(error) if attempt < 4 && is_transient(&error) => {
                        tokio::time::sleep(Duration::from_millis(1 << attempt)).await;
                    }
                    Err(error) => return Err(map_error(error)),
                }
            }
            Err(conflict())
        })
    }
}

async fn allocate_once(
    pool: &Pool,
    request: &SequenceAllocationRequest,
    machine_id: &[u8],
) -> Result<u64, MySqlError> {
    let mut connection = pool.get_conn().await?;
    let row: Option<(u64,)> = connection
        .exec_first(
            "CALL identifold_allocate_sequence(\
             :machine_id, :namespace, :reference_prefix, :scope, :width)",
            params! {
                "machine_id" => machine_id,
                "namespace" => request.namespace.as_str(),
                "reference_prefix" => request.reference_prefix.as_str(),
                "scope" => request.scope.as_deref(),
                "width" => request.width,
            },
        )
        .await?;
    row.map(|value| value.0)
        .ok_or_else(|| MySqlError::Other(Box::new(StorageResultError)))
}

#[derive(Debug)]
struct StorageResultError;

impl std::fmt::Display for StorageResultError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        formatter.write_str("invalid storage result")
    }
}

impl std::error::Error for StorageResultError {}

fn mapping(machine_id: Vec<u8>, namespace: String) -> Result<ReferenceMapping, Error> {
    Ok(ReferenceMapping {
        machine_id: machine_id_string(&machine_id)?,
        namespace,
    })
}

fn machine_id_bytes(value: &str) -> Result<Vec<u8>, Error> {
    let hexadecimal: String = value.chars().filter(|value| *value != '-').collect();
    if hexadecimal.len() != 32 || !hexadecimal.bytes().all(|value| value.is_ascii_hexdigit()) {
        return Err(conflict());
    }
    (0..32)
        .step_by(2)
        .map(|index| u8::from_str_radix(&hexadecimal[index..index + 2], 16).map_err(|_| conflict()))
        .collect()
}

fn machine_id_string(value: &[u8]) -> Result<String, Error> {
    if value.len() != 16 {
        return Err(conflict());
    }
    let hexadecimal: String = value.iter().map(|byte| format!("{byte:02x}")).collect();
    Ok(format!(
        "{}-{}-{}-{}-{}",
        &hexadecimal[0..8],
        &hexadecimal[8..12],
        &hexadecimal[12..16],
        &hexadecimal[16..20],
        &hexadecimal[20..32]
    ))
}

fn parse_sequential_reference(reference: &str) -> Option<(&str, &str, &str)> {
    let parts: Vec<_> = reference.split('-').collect();
    let (prefix, scope, sequence, check) = match parts.as_slice() {
        [prefix, sequence, check] => (*prefix, "", *sequence, *check),
        [prefix, scope, sequence, check] => (*prefix, *scope, *sequence, *check),
        _ => return None,
    };
    if !(2..=8).contains(&prefix.len())
        || !prefix.bytes().all(|value| value.is_ascii_uppercase())
        || (!scope.is_empty()
            && (scope.len() != 4 || !scope.bytes().all(|value| value.is_ascii_digit())))
        || !(4..=18).contains(&sequence.len())
        || !sequence.bytes().all(|value| value.is_ascii_digit())
        || check.len() != 1
    {
        return None;
    }
    Some((prefix, scope, sequence))
}

fn map_error(error: MySqlError) -> Error {
    match error {
        MySqlError::Server(error) if error.state == "22003" => Error("sequence_overflow"),
        MySqlError::Server(error) if error.state == "22023" => Error("invalid_allocation_policy"),
        _ => conflict(),
    }
}

fn is_transient(error: &MySqlError) -> bool {
    matches!(
        error,
        MySqlError::Server(error)
            if error.code == 1205 || error.code == 1213 || error.state == "40001"
    )
}

fn conflict() -> Error {
    Error("allocation_conflict")
}
